src/test/template-analyzer_test.ts

```diff
@@ -0,0 +1,502 @@
+import * as assert from 'assert';
+import treeAdapter = require('parse5-htmlparser2-tree-adapter');
+import {TemplateAnalyzer} from '../template-analyzer';
+import {parse} from 'babel-eslint';
+import {TaggedTemplateExpression} from 'estree';
+
+/**
+ * Helper for getting the TaggedTemplateExpression from a string of code
+ * @param {string} code - code to parse
+ * @return {TaggedTemplateExpression|null}
+ */
+function getTaggedTemplateExpression(
+  code: string
+): TaggedTemplateExpression | null {
+  const ast = parse(code);
+  const statement = ast.body[0];
+  if (
+    statement.type !== 'ExpressionStatement' ||
+    statement.expression.type !== 'TaggedTemplateExpression'
+  ) {
+    return null;
+  }
+  return statement.expression;
+}
+
+describe('TemplateAnalyzer', () => {
+  describe('create', () => {
+    it('should reuse the analyzer for a given TaggedTemplateExpression', () => {
+      const taggedTemplate = getTaggedTemplateExpression(
+        'html`<h1>Hello</h1>`'
+      )!;
+      const analyzer1 = TemplateAnalyzer.create(taggedTemplate);
+      const analyzer2 = TemplateAnalyzer.create(taggedTemplate);
+      assert.strictEqual(analyzer1, analyzer2);
+    });
+    it('should handle parse errors', () => {
+      const taggedTemplate = getTaggedTemplateExpression(
+        'html`<h1>Hello<///h1>>>>`'
+      )!;
+      const analyzer = TemplateAnalyzer.create(taggedTemplate);
+      assert.deepEqual(analyzer.errors, [
+        {
+          code: 'invalid-first-character-of-tag-name',
+          endCol: 12,
+          endLine: 1,
+          endOffset: 11,
+          startCol: 12,
+          startLine: 1,
+          startOffset: 11
+        }
+      ]);
+    });
+  });
+  describe('getLocationFor', () => {
+    /* eslint-disable-next-line max-len */
+    it('should return the location of the starting tag of an element', (done) => {
+      const taggedTemplate = getTaggedTemplateExpression(
+        'html`<h1>Hello</h1>`'
+      )!;
+      const analyzer = TemplateAnalyzer.create(taggedTemplate);
+      let actual = 0;
+
+      analyzer.traverse({
```

Comment:
instead of having assertions inside a callback (and risking them never being called), can you make `_ast` of the analyzer public (and rename it to `ast`) then cast the correct child to a `treeAdapter.Element` and pass it directly into `getLocationFor`?

sounds like it should work.. possibly

Reply:
also, you can then add tests for the undefined location branches instead of ignoring them.

IIRC the type of a node has `sourceCodeLocation?: Location`. So you should make a node, set `sourceCodeLocation = undefined`, then pass it to `getLocationFor` and cover those branches

if we must see a 100% even through ignoring, at least then the ignore flags only exist in the rules where we can't set source code locations so easily.

Reply:
Sorry, I'm unsure what you mean by the first comment... The second comment goes back to our discussion about location checks. In eslint there are no cases where we won't have a location, so I'm unsure what adding a test case/including it in code coverage would do for us.

Reply:
the types define that it can be undefined. i would rather we test that case even if we are likely to never encounter it, than scatter coverage flags through the files to trick ourselves into thinking we covered everything.

if at some point you or i contribute to the types to make the locations non-nullable, you are right and we can delete the test cases.

should be a super simple test: get a node, set the location to `undefined`, pass it in.

re my first comment, you're traversing the AST in this test for sole purpose of getting hold of a parsed node. this means if the traversal mechanism were to ever fail, the assertion may not get hit (as its in a callback) and we wouldn't notice as tests would still pass.

instead, we should make `_ast` a public property and access it directly in these particular tests. so instead of traversing to get the node we want, we instead `(analyzer.ast.foo as treeAdapter.Whatever).bar as treeAdapter.AnotherThing;` and pass that into our analyzer method directly

Reply:
the test would become:

```ts
const node = analyzer.ast.foo.bar;
assert.deepEqual(analyzer.getLocationFor(node), {
  ...
});
```